config.go

```diff
@@ -18,18 +70,79 @@ type Config struct {
 	InvalidUsernameMessage string   `json:"invalid-username-message"`
 }
 
+type ProxyPolicyConfig struct {
+	AllowedCIDRs []netip.Prefix
+	AllowedHosts []string
+}
+
 type UsernamePolicyConfig struct {
-	InvalidUsername        []string `json:"invalid-username"`
-	InvalidUsernameMessage string   `json:"invalid-username-message"`
+	InvalidUsernames       []string
+	InvalidUsernameMessage string
 }
 
 type PasswordPolicyConfig struct {
-	AllUsernameNoPassword bool     `json:"all-username-nopassword"`
-	UsernameNoPassword    []string `json:"username-nopassword"`
+	AllUsernameNoPassword bool
+	UsernamesNoPassword   []string
 }
 
-type RecoveryConfig struct {
-	Server   string   `json:"recovery-server"`
-	Username []string `json:"recovery-username"`
-	Token    string   `json:"token"`
+func convertLegacyConfig(config LegacyConfig) Config {
+	if config.RecoveryToken == "" {
+		config.RecoveryToken = config.Token
+	}
+	hostKeys := make([]SSHKeyConfig, 0, len(config.HostKeys))
+	for _, path := range config.HostKeys {
+		hostKeys = append(hostKeys, SSHKeyConfig{Path: path})
+	}
+	return Config{
+		Address: config.Address,
+		SSH: SSHConfig{
+			Banner:   config.Banner,
+			HostKeys: hostKeys,
+		},
+		Auth: AuthConfig{
+			Endpoint:               config.API,
+			Token:                  config.Token,
+			InvalidUsernames:       config.InvalidUsername,
+			InvalidUsernameMessage: config.InvalidUsernameMessage,
+			AllUsernameNoPassword:  config.AllUsernameNoPassword,
+			UsernamesNoPassword:    config.UsernameNoPassword,
+		},
+		Logger: LoggerConfig{
+			Enabled:  config.Logger != "",
+			Endpoint: fmt.Sprintf("udp://%s", config.Logger),
+		},
+		ProxyProtocol: ProxyProtocolConfig{
+			Enabled:  len(config.ProxyCIDRs) > 0,
+			Networks: config.ProxyCIDRs,
+		},
+		Recovery: RecoveryConfig{
+			Address:   config.RecoveryServer,
+			Usernames: config.RecoveryUsername,
+			Token:     config.RecoveryToken,
+		},
+	}
+}
+
+func convertProxyPolicyConfig(config ProxyProtocolConfig) (*ProxyPolicyConfig, error) {
+	if config.Enabled != true {
+		return &ProxyPolicyConfig{}, nil
```

Comment:
This further invalidates the utility of returning pointers.

Reply:
I'm not sure how the distinction between "empty" and "invalid" would be useful. In either case we can just disable the feature.

Reply:
We’re failing the server initialization if the config is invalid. This helps for static checking on the configuration file, and for a stable and predictable behavior.

Reply:
I'm still confused. We're already returning `(something, error)` following Go conventions. An invalid config should be signified by a non-`nil` error value.

Also, it would be great if `proxy.enabled`, as an optional feature, defaults to false instead of being mandatory (as stated in README).

Reply:
`logger.enabled` and `proxy-protocol.enabled` are really optional, but I didn't find a way to nicely document their default value. It doesn't look so nice if we add a dedicated "Default" column.

Reply:
> I'm still confused. We're already returning (something, error) following Go conventions. An invalid config should be signified by a non-nil error value.

Makes sense, changes are all set.